File: examples/language-samples/sample.nim
# Real Nim: procs, objects, iterators, templates, pragmas.
import std/[strutils, sequtils, tables]

type
  Status* = enum
    Draft, Sent, Paid

  Invoice* = ref object of RootObj
    id*: int
    customer*: string
    amount*: float
    status*: Status

proc newInvoice*(id: int, amount: float = 0.0): Invoice =
  result = Invoice(id: id, amount: amount, status: Draft)

method describe*(self: Invoice): string {.base.} =
  if self.amount > 10_000:
    "large"
  elif self.amount > 0:
    "standard"
  else:
    "empty"

iterator overdue*(invoices: seq[Invoice]): Invoice =
  for invoice in invoices:
    if invoice.status == Sent and invoice.amount > 0:
      yield invoice

template withCache(body: untyped): untyped =
  var cache {.inject.} = initTable[int, Invoice]()
  body

when isMainModule:
  let items = @[newInvoice(1, 42.0), newInvoice(2, 20000.0)]
  echo items.mapIt(it.describe).join(", ")

File: examples/language-samples/sample.zig
// Real Zig: comptime, error unions, allocators, slices.
const std = @import("std");
const Allocator = std.mem.Allocator;

const InvoiceError = error{
    NotFound,
    InvalidAmount,
};

pub const Status = enum { draft, sent, paid };

pub const Invoice = struct {
    id: u32,
    amount: f64 = 0.0,
    status: Status = .draft,

    pub fn describe(self: Invoice) []const u8 {
        return if (self.amount > 10_000) "large" else if (self.amount > 0) "standard" else "empty";
    }
};

pub fn total(invoices: []const Invoice) f64 {
    var sum: f64 = 0;
    for (invoices) |invoice| {
        sum += invoice.amount;
    }
    return sum;
}

pub fn find(list: []const Invoice, id: u32) InvoiceError!Invoice {
    for (list) |invoice| {
        if (invoice.id == id) return invoice;
    }
    return InvoiceError.NotFound;
}

test "total sums amounts" {
    const items = [_]Invoice{ .{ .id = 1, .amount = 10 }, .{ .id = 2, .amount = 5 } };
    try std.testing.expectEqual(@as(f64, 15), total(&items));
}
